src/vstest.console/TestPlatformHelpers/TestRequestManager.cs

```diff
@@ -155,43 +155,45 @@ public bool DiscoverTests(DiscoveryRequestPayload discoveryPayload, ITestDiscove
             var criteria = new DiscoveryCriteria(discoveryPayload.Sources, batchSize, this.commandLineOptions.TestStatsEventTimeout, runsettings);
             criteria.TestCaseFilter = this.commandLineOptions.TestCaseFilterValue;
 
-            using (IDiscoveryRequest discoveryRequest = this.testPlatform.CreateDiscoveryRequest(requestData, criteria))
+            try
             {
-                try
+                using (IDiscoveryRequest discoveryRequest = this.testPlatform.CreateDiscoveryRequest(requestData, criteria))
                 {
-                    this.testLoggerManager?.RegisterDiscoveryEvents(discoveryRequest);
-                    discoveryEventsRegistrar?.RegisterDiscoveryEvents(discoveryRequest);
+                    try
+                    {
+                        this.testLoggerManager?.RegisterDiscoveryEvents(discoveryRequest);
+                        discoveryEventsRegistrar?.RegisterDiscoveryEvents(discoveryRequest);
 
-                    this.testPlatformEventSource.DiscoveryRequestStart();
+                        this.testPlatformEventSource.DiscoveryRequestStart();
 
-                    discoveryRequest.DiscoverAsync();
-                    discoveryRequest.WaitForCompletion();
+                        discoveryRequest.DiscoverAsync();
+                        discoveryRequest.WaitForCompletion();
 
-                    success = true;
-                }
-                catch (Exception ex)
-                {
-                    if (ex is TestPlatformException ||
-                        ex is SettingsException ||
-                        ex is InvalidOperationException)
-                    {
-#if TODO
-                        Utilities.RaiseTestRunError(testLoggerManager, null, ex);
-#endif
-                        success = false;
+                        success = true;
                     }
-                    else
+
+                    finally
                     {
-                        throw;
+                        this.testLoggerManager?.UnregisterDiscoveryEvents(discoveryRequest);
+                        discoveryEventsRegistrar?.UnregisterDiscoveryEvents(discoveryRequest);
                     }
                 }
-                finally
+            }
+            catch (Exception ex)
+            {
+                if (ex is TestPlatformException ||
```

Comment:
nit : ` catch (Exception ex) when ( ex is TestPlatformException || ex is SettingsException) `